### PR TITLE
feat(trigger/pubsub): Suggest subscriptions from echo configuration.

### PR DESCRIPTION
**Pull request by jtk54**

### CODE REVIEW

File: app/scripts/modules/core/src/pubsub/pubsubSubscription.service.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { module } from 'angular';
+
+import { API_SERVICE, Api } from 'core/api/api.service';
+
+export class PubsubSubscriptionService {
+  constructor(private API: Api) { 'ngInject'; }
+
+  public getPubsubSubscriptions(): ng.IPromise<string[]> {
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pubsub/pubsubSubscription.service.ts]: You should drop the `ng` namespacing by importing `IPromise` from `angular`.

File: app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts [hunk 2]
```diff
@@ -1,19 +1,29 @@
...
 
 class PubsubTriggerController implements IController {
   public pubsubSystems = ['kafka', 'google'];
+  public pubsubSubscriptions: string[];
 
-  constructor(public trigger: IPubsubTrigger) {
+  constructor(public trigger: IPubsubTrigger,
+              public pubsubSubscriptionService: PubsubSubscriptionService) {
     'ngInject';
+    pubsubSubscriptionService.getPubsubSubscriptions().then(subscriptions => {
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts]: Ideally the full set of states of this call would be surfaced to the user. We generally don't end up surfacing failure, but you should show a spinner that indicates that the subscriptions are still loading. See https://github.com/spinnaker/deck/blob/master/app/scripts/modules/core/src/pipeline/config/triggers/jenkins/jenkinsTrigger.module.js#L89-L98

For a failed call, you can add `.catch(() => this.pubsubSubscriptions = []);`

File: app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts [hunk 3]
```diff
@@ -1,19 +1,29 @@
...
+  PubsubSubscriptionService,
+} from '@spinnaker/core';
 
 class PubsubTriggerController implements IController {
   public pubsubSystems = ['kafka', 'google'];
+  public pubsubSubscriptions: string[];
 
-  constructor(public trigger: IPubsubTrigger) {
+  constructor(public trigger: IPubsubTrigger,
+              public pubsubSubscriptionService: PubsubSubscriptionService) {
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts]: Should be `private`.

File: app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts [hunk 4]
```diff
@@ -1,19 +1,29 @@
 import { IController, module } from 'angular';
 
 import { PIPELINE_CONFIG_PROVIDER, PipelineConfigProvider } from 'core/pipeline/config/pipelineConfigProvider';
-import { IPubsubTrigger } from '@spinnaker/core';
+import {
+  IPubsubTrigger,
+  PUBSUB_SUBSCRIPTION_SERVICE,
+  PubsubSubscriptionService,
+} from '@spinnaker/core';
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts]: Don't use `@spinnaker/core` from inside `core`. Use `core/pubsub` instead.

Comment (jtk54) [reply app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts]: Done.

File: app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts (hunk 3 above)
Comment (jtk54) [reply app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts]: `private` throws a linter error, I just removed `public`.

File: app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts (hunk 2 above)
Comment (jtk54) [reply app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts]: Done.

File: app/scripts/modules/core/src/pubsub/pubsubSubscription.service.ts (hunk 1 above)
Comment (jtk54) [reply app/scripts/modules/core/src/pubsub/pubsubSubscription.service.ts]: Done.

File: app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts [hunk 5]
```diff
@@ -1,19 +1,35 @@
...
-  constructor(public trigger: IPubsubTrigger) {
+  constructor(public trigger: IPubsubTrigger,
+              pubsubSubscriptionService: PubsubSubscriptionService) {
     'ngInject';
+
+    this.subscriptionsLoaded = false;
+    pubsubSubscriptionService.getPubsubSubscriptions().then(subscriptions => {
+      this.subscriptionsLoaded = true;
+      this.pubsubSubscriptions = subscriptions;
+    }).catch(() => this.pubsubSubscriptions = []);
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/triggers/pubsub/pubsub.trigger.ts]: Should flip `subscriptionsLoaded` even in the case where the call failed (the spinner should stop spinning and you should see an empty dropdown).

You can use `.finally` here:

```
pubsubSubscriptionService.getPubsubSubscriptions()
  .then(subscriptions => this.pubsubSubscriptions = subscriptions)
  .catch(() => this.subscriptions = [])
  .finally(() => this.subscriptionsLoaded = true);
```

**Pull request merged**